File: 83_Remove_Duplicates_from_Sorted_List.js
/**
 * Given a sorted linked list, delete all duplicates such that each element appear only once.
 * For example,
 * Given 1->1->2, return 1->2.
 * Given 1->1->2->3->3, return 1->2->3.
 */

/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var deleteDuplicates = function(head) {
    if (head === null) return head;
    
    var p = head;
    while (p !== null) {
        while (p.next !== null && p.val === p.next.val) {
            p.next = p.next.next;
        }
        p = p.next;
    }

    return head;
};

File: 83_Remove_Duplicates_from_Sorted_List.c
#include <stdlib.h>

struct ListNode {
  int val;
  struct ListNode *next;
};

struct ListNode* deleteDuplicates(struct ListNode* head) {
  if (!head || !head->next) return head;

  struct ListNode *current = head;
  while (current->next != NULL) {
    if (current->val == current->next->val) {
      current->next = current->next->next;
    } else {
      current = current->next;
    }
  }

  return head;
}

int main() {
  struct ListNode *test_2 = {2, NULL};
  struct ListNode *test_1 = {1, test_2};
  // struct ListNode *test_1_1 = {1, test_1};
  struct ListNode *test_0 = {0, test_1};

  deleteDuplicates(test_0);

}
